lib/store_base_sti_class_for_5_0.rb

```diff
@@ -0,0 +1,401 @@
+            end
+
+            if preload_values[:readonly] || values[:readonly]
+              scope.readonly!
+            end
+
+            if options[:as]
+              # START PATCH
+              # original:
+              # scope.where!(klass.table_name => { reflection.type => model.base_class.sti_name })
+
+              scope.where!(klass.table_name => { reflection.type => ActiveRecord::Base.store_base_sti_class ? model.base_class.sti_name : model.sti_name })
+
+              # END PATCH
+            end
+
+            scope.unscope_values = Array(values[:unscope]) + Array(preload_values[:unscope])
+            klass.default_scoped.merge(scope)
+          end
+        end
+
+        module ThroughAssociation
+          private
+
+          def through_scope
+            scope = through_reflection.klass.unscoped
+
+            if options[:source_type]
+              # BEGIN PATCH
+              # original: scope.where! reflection.foreign_type => options[:source_type]
+
+              adjusted_foreign_type =
+                if ActiveRecord::Base.store_base_sti_class
+                  options[:source_type]
+                else
+                  ([options[:source_type].constantize] + options[:source_type].constantize.descendants).map(&:to_s)
+                end
+
+              scope.where! reflection.foreign_type => adjusted_foreign_type
+
+              # END PATCH
+            else
+              unless reflection_scope.where_clause.empty?
+                scope.includes_values = Array(reflection_scope.values[:includes] || options[:source])
+                scope.where_clause = reflection_scope.where_clause
+              end
+
+              scope.references! reflection_scope.values[:references]
+              if scope.eager_loading? && order_values = reflection_scope.values[:order]
+                scope = scope.order(order_values)
+              end
+            end
+
+            scope
+          end
+        end
+      end
+
+      class AssociationScope
+
+        def self.get_bind_values(owner, chain)
+          binds = []
+          last_reflection = chain.last
+
+          binds << last_reflection.join_id_for(owner)
+          if last_reflection.type
+            # START PATCH
+            # original: binds << owner.class.base_class.name
+            binds << (ActiveRecord::Base.store_base_sti_class ? owner.class.base_class.name : owner.class.name)
+            # END PATCH
+          end
+
+          chain.each_cons(2).each do |reflection, next_reflection|
+            if reflection.type
+              # START PATCH
+              # original: binds << next_reflection.klass.base_class.name
+              binds << (ActiveRecord::Base.store_base_sti_class ? next_reflection.klass.base_class.name : next_reflection.klass.name)
+              # END PATCH
+            end
+          end
+          binds
+        end
+
+
+        def next_chain_scope(scope, table, reflection, association_klass, foreign_table, next_reflection)
```

Comment:
This method is private in the original implementation. Also, there's an extra newline above.